File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        
        // maintain an vector with a leader and margin of error
        int left = 0;
        int right = 1;
        int fin_max = 0;
        
        // initialize vector
        vector<int> c_window(26, 0);
        c_window[s[0]-65]++;
        c_window[s[1]-65]++;
                
        while(right < s.size()) {
            int count = 0;
            int leader = 0;
            // count up character replacement
            for(int i = 0; i < c_window.size(); i++) {
                count += c_window[i];
                leader = max(leader, c_window[i]);
            }
            if(count - leader > k) {
                c_window[s[left]-65]--;
                left++;
            }
            else {
                fin_max = max(right - left + 1, fin_max);
                right++;
                if(right < s.size()) {
                    c_window[s[right]-65]++;
                }
            }
        }
        
        return fin_max;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        
        l = 0
        r = 0
        longest = 0
        window = 26 * [0]
        window[ord(s[0])-65] += 1


        while r < len(s):            
            # find the current leader
            leader_count = 0
            for i in range(len(window)):
                if window[i] > leader_count:
                    leader_count = window[i]
                    
            window_length = r-l+1
            if window_length > k + leader_count:
                window[ord(s[l])-65] -= 1
                l += 1
            else:
                longest = max(longest, window_length)
                r += 1
                if r < len(s):
                    window[ord(s[r])-65] += 1
        return longest

            
            
            
        
        